Add `StrokeCap` to `CircularProgressIndicator`

bernaferrari opened the pull request:
Sequel to https://github.com/flutter/flutter/pull/109111
Fix https://github.com/flutter/flutter/issues/109137

![image](https://user-images.githubusercontent.com/351125/225201457-716f7263-cbb1-4d22-95ba-63a254dbc95a.png)

~I can make `LinearProgressIndicator` accept BorderRadius (instead of round/square), but I can't (and no one can) make `CircularProgressIndicator` work with that. You would need someone from Dart/Skia team that is good at math to make drawArc more flexible.~

~So I used strokeCap everywhere, even though borderRadius (`LinearProgressIndicator`) + strokeCap (`CircularProgressIndicator`) would be better IMO. In a future PR I think I could add borderRadius to the "background" of the LinearProgressIndicator, so a borderRadius to the foreground could make a lot of sense. But I think people would be happy already with the round option. What do you think? Option 1 or 2?~

Disclaimer: I don't think StrokeCap works that well for what we are doing. It has 3 options, butt, square and round, but 99% of the time you want two of them, and you never know which two you want. `CircularProgressIndicator` uses square for indeterminate (paints a bit more, so it is always visible, even on 0), butt for determinate (paints exactly). `CircularProgressIndicator` doesn't use it.

Something like `isIndicatorRound` would be better, since all we want is true/false. Do you agree?

```dart
CircularProgressIndicator(
  strokeCap: StrokeCap.round,
),
// Option 1. This PR.
LinearProgressIndicator(
  strokeCap: StrokeCap.round,
),
```
or
```dart
// Option 2. Different proposal. LinearProgressIndicator only.
LinearProgressIndicator(
  backgroundRadius: BorderRadius.circular(8),
  indicatorRadius: BorderRadius.circular(8),
),
```

cc @Piinks

Hari-07 committed:
Added strokecap option on progress on circular/refresh progress indicator

Fixed typo in existing test

Added ability to mock strokecap

Wrote test for Refresh Progress Indicator

Piinks reviewed (commented):
There are some failing tests, can you take a look?
Also, I'd like @Hari-07 to weigh in here. They originally authored the PR and stated they were already re-staging it.

Hari-07 commented:
I did rebase my changes but since it'd be identical to this, I think I'll refrain from opening a duplicate PR. As for the discussions about boolean or StrokeCap, I think exposing it as a boolean makes sense, and then setting it to the correct strokecap inside it

The original issue #71085 which got you to reopen #109137, mentions customizing the radius. From what I know the `drawArc` method at the lowest level interfaces with skia functions, and those read off the `StrokeCap` on the `Paint` instance used to do the rendering calculation so in order to allow that radius to be customized I believe a new property would have to be added on `Paint` , and I think that'd have to be updated in the engine repo

Thanks for taking my input @Piinks

bernaferrari commented:
I just changed `strokeCap` to `hasRoundStroke` boolean property. I think the test fail was a flake. Let's see now.

whiskeyPeak commented:
Since the end goal is to potentially add a border radius to both widgets, I think it would make a bit more sense if LinearProgresseIndicator is given the customizable borderRadius property now.

bernaferrari commented:
I don't think adding borderRadius to CircularProgress would ever be possible, but we can decide what to do with the Linear one because it is quite different. I'm just afraid of polluting the theme too much.

bleroux commented on packages/flutter/test/rendering/mock_canvas.dart:
```diff
@@ -940,6 +942,9 @@ abstract class _DrawCommandPaintPredicate extends _PaintPredicate {
     if (shader != null && !shader!.matches(paintArgument.shader, <dynamic, dynamic>{})) {
       throw 'It called $methodName with a paint whose shader, ${paintArgument.shader}, was not exactly the expected shader ($shader).';
     }
+    if(strokeCap != null && paintArgument.strokeCap != strokeCap) {
```
nit: missing space after `if`

bernaferrari replied:
Sharp eyes, thanks!

bleroux commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -108,6 +109,20 @@ abstract class ProgressIndicator extends StatefulWidget {
+  /// stroke.
+  ///
+  /// If [ProgressIndicator.hasRoundStroke] is null then it will use the
+  /// ambient [ProgressIndicatorThemeData.hasRoundStroke]. If that is null
```
Consider adding a test to check that ProgressIndicatorThemeData fallback work as expected.

bernaferrari replied:
Thanks. Fixed.

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -49,6 +49,7 @@ abstract class ProgressIndicator extends StatefulWidget {
     this.valueColor,
     this.semanticsLabel,
     this.semanticsValue,
+    this.hasRoundStroke,
```
Since StrokeCap.round is only relevant to CircularProgressIndicator (If the Linear one chooses a radius instead), should this only be on the CircularProgressIndicator class?

Piinks replied:
I am also not sure if the name is clear as a user, why not use StrokeCap directly?

bernaferrari replied:
StrokeCap has 3 values and only 2 of them are relevant. Butt and Square collide with each other, because Circular uses Butt when it has value, Square when it doesn't (it is spinning).

bernaferrari replied:
> should this only be on the CircularProgressIndicator

Probably. How do you think the theme should be? `CircularHasRoundIndicator` and `LinearIndicatorRadius`?

bernaferrari replied:
Renamed to `preferRoundIndicator`.

Piinks replied:
> only 2 of them are relevant

I disagree with this statement, why are there 3 then? :) It's not for us to decide that one is irrelevant, the uses cases have been pointed out in the discussion of this PR.

> How do you think the theme should be? CircularHasRoundIndicator and LinearIndicatorRadius?

Are there not individual component themes for circular and linear? I would not put both in the super class ProgressIndicatorTheme.

bernaferrari replied:
> I disagree with this statement, why are there 3 then? :) It's not for us to decide that one is irrelevant, the uses cases have been pointed out in the discussion of this PR.

If it is square, the `value` is represented wrong: https://github.com/flutter/flutter/issues/109137#issuecomment-1209391300.
If it is butt, it is represented correctly.

But if it is butt, on indeterminate, it "disappears", so whoever implemented it made it square so it is always visible.

Exposing this to the user is hard: on square they will have the wrong value, on butt they will have an "unintended" animation (which, if it is to be customizable, we could allow changing other things, such as speed).

> Are there not individual component themes for circular and linear?

No, I could make one, I guess..

Piinks replied:
> No, I could make one, I guess..

Oh, I meant that if they don't exist then we should probably omit theming support for now. :) No worries, no need to create more themes right now for this change.

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -145,13 +166,15 @@ class _LinearProgressIndicatorPainter extends CustomPainter {
     this.value,
     required this.animationValue,
     required this.textDirection,
+    required this.hasRoundStroke,
```
I think having a radius for the linear one makes sense. Also, which stroke? The value one? Or the track?

bernaferrari replied:
`hasRoundIndicator`?

I'll change linear, just tell me what the name should be for the theme to work...

Piinks replied:
See comment above on theming

bernaferrari committed:
Improve PR

Fix tests by removing them, for now.


Use boolean instead.

Fix PR review.

Rename to `preferRoundIndicator`.

bernaferrari commented:
With `LinearProgressIndicator` + borderRadius, I get this:

<img width="73" alt="image" src="https://user-images.githubusercontent.com/351125/226039441-a82d1dd0-981a-4c71-b99e-157809efde9d.png">

The logic I did is:

```dart
final BorderRadius backgroundBorderRadius = widget.backgroundBorderRadius == null && widget.value != BorderRadius.zero // could also be != null and then it would be nullable
        ? widget.indicatorBorderRadius
        : BorderRadius.zero;
```

What do you think? Nullable or Zero?

bernaferrari commented:
How do we call the `borderRadius`? `indicatorBorderRadius` and `borderRadius` (1) or `backgroundBorderRadius` (2)? I'm tempted to go with 1, because it applies to the whole shape, not the background only.

<img width="474" alt="image" src="https://user-images.githubusercontent.com/351125/226083798-71e529e9-4ae9-41de-a500-0412b989a273.png">

The only "unusual" thing I made is that the track has Radius.zero in the start position (when it it is determined). This prevents the left side from being round, which doesn't make sense.

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -303,6 +333,14 @@ class LinearProgressIndicator extends ProgressIndicator {
   final double? minHeight;
 
+  /// The border radius of the indicator.
+  /// By default it is [BorderRadius.zero], which produces a rectangular indicator.
```
```suggestion
  /// 
  /// By default it is [BorderRadius.zero], which produces a rectangular indicator.
```

bernaferrari replied:
Thanks!

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -303,6 +333,14 @@ class LinearProgressIndicator extends ProgressIndicator {
+  final BorderRadius indicatorBorderRadius;
+
+  /// The border radius of the background indicator.
+  /// By default it is null, which produces a rectangular track.
```
```suggestion
  /// 
  /// By default it is null, which produces a rectangular track.
```
This is a non-nullable parameter. How does it default to null? :)

bernaferrari replied:
ooops :P

bernaferrari replied:
Actually, I was playing with a few things and forgot to update the documentation. I was checking if I could "auto add a borderRadius" but my idea was bad and it is better to allow the user to break things. IndicatorBorderRadius = 100 + background = 0 result in a weird scenario I was trying to avoid, but I'm letting the user decide.

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -354,6 +392,15 @@ class _LinearProgressIndicatorState extends State<LinearProgressIndicator> with
       context: context,
       child: Container(
+        clipBehavior: widget.backgroundBorderRadius != BorderRadius.zero
+            ? Clip.antiAlias
+            : Clip.none,
```
```suggestion
          ? Clip.antiAlias
          : Clip.none,
```

bernaferrari replied:
Thanks!

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -544,6 +604,14 @@ class CircularProgressIndicator extends ProgressIndicator {
   /// The width of the line used to draw the circle.
   final double strokeWidth;
 
+  /// The progress indicator's line ending: square or round.
```
> square or round

The name/boolean/square-round is kind of confusing here.

I think StrokeCap is still just the straightforward solution.

bernaferrari replied:
The issue with strokeCap, as I tried to explain before, is that butt/square are always wrong. If you want butt, indeterminate is wrong. If you want square, value is wrong. Would we have a fourth value (null?) that goes between butt and square? I think that's too much complexity for the user. Even introducing a new enum could be easier to understand.

Piinks replied:
Is StrokeCap not properly documented? Again, these are decisions I do not think are fair for us to make for the developer. I think this should be StrokeCap.

bernaferrari replied:
How do you see the API working? Because either value (square/butt) would be breaking. If they are overridden, they don't matter because they are the same value.

Piinks replied:
How would it be breaking? Does exposing this value to developers break the default behavior?

Piinks replied:
Can the value of CircularProgressIndicator.strokeCap be nullable? in which case it would choose this default behavior when the developer has not provided a value?

bernaferrari replied:
I can, but just to double check, you acknowledge this might be error prone, because no one would want a value that is represented wrongly (https://github.com/flutter/flutter/issues/109137#issuecomment-1209391300). For the "different indeterminate animation" I think it is fine (but different enough than the official material animation).

Piinks replied:
> you acknowledge this might be error prone

I don't. I only pointed out that this value can be exposed without introducing a breaking change and suggested a way to do so. :)

bernaferrari replied:
Oh, I totally agree. The weird thing for me is that there would be 4 values and only 2 would be really "correct". So even though we want to give users options they will shoot themselves with 2 of them (which seem the correct thing, you have `circle`, you want... `square`, bang).

bernaferrari replied:
If the progress value starts on -10 instead of 0, we have a "deterministic" problem. No one would want it to start before, and if they want, we could just expose that number. Same thing, if they want value: 40, they don't want value: 50 (else we would expose like an offset and let they play with it).

For indeterminate, it is less bad because the material animation would change (most of them wouldn't notice), so I believe, if we want the material animation to be customizable, there are other ways of doing this (like allowing to change the speed, curves and delay).

bleroux replied:
Just my 2cts, I think it is fair to assume that a determinate `CircularIndicator` is "correct" if it starts right at 90 degrees and that it draws half of a circle when its value is 0.5. Both are not true when using `StrokeCap.square` (not to say it is not 'circular' because ‘StrokeCap.square’ really means there are two squares drawn at both ends of the arc, which might makes sense if you want to ‘connect’ lines and arcs).

Nonetheless, @Pinks point about the versatility of the public API is valid, especially considering @bernaferrari following argument: “one day they might add other strokeCap effects, it makes sense.” (maybe Impeller will).

bernaferrari replied:
I doubt so because CSS is 20 years old and only supports the same 3 strokeCap. But I agree they could. I just don't know what it could be that would add benefit to the framework.

Piinks commented on packages/flutter/test/material/progress_indicator_test.dart:
```diff
@@ -443,6 +443,45 @@ void main() {
+    expect(
+        find.byType(LinearProgressIndicator),
+        paints
+        ..rrect(
+          rrect: RRect.fromLTRBAndCorners(0.0, 0.0, 25.0, 4.0,
+            topRight: const Radius.circular(10.0),
+            bottomRight: const Radius.circular(10.0)),
```
```suggestion
          ..rrect(
            rrect: RRect.fromLTRBAndCorners(0.0, 0.0, 25.0, 4.0,
              topRight: const Radius.circular(10.0),
              bottomRight: const Radius.circular(10.0),
            ),
```

bernaferrari commented:
I'll split this in two, so we can discuss the round thing a bit more (I have an idea) while merging the other part. I think it will be easier and less code to deal with.

Piinks commented:
> (I have an idea)

Would you like to share? :)

bernaferrari commented:
Oh, I completely forgot. We could have `strokeCap`, make the default `square` (because sounds more logical) and we override it on `value` to butt. So there is no way users would get a "wrong" value. For butt + indeterminate, the animation would get weird, but that would be fine.

Therefore, square + round could live together as the two options everybody would use.

Piinks commented:
> So there is no way users would get a "wrong" value

I still do not understand what a wrong value is. What if the user wants that value?

bernaferrari commented:
I'm not sure you saw this comment, but I fully agree. The user would never expect this value because it breaks the basic assumption of a value. If we want this to be exposed, we could have like `startAngle`.

<img width="561" alt="image" src="https://user-images.githubusercontent.com/351125/229938824-907b2eee-4a38-41ea-a8a2-c1f119e1b7ff.png">

Piinks commented:
> The user would never expect this value

I do not follow. What value?

bernaferrari commented:
For example, you don't expect 0.95 to be 1.0, or 0.5 to not be half:

![Frame 30](https://user-images.githubusercontent.com/351125/229939771-9de5ded6-5b17-4237-a111-0cf6c9c69ab5.png)

Piinks commented:
This is documented behavior: https://api.flutter.dev/flutter/dart-ui/StrokeCap.html

bernaferrari commented:
I agree, but for the ProgressIndicator it doesn't work, which is why the person that implemented it made square for value, butt for indeterminate. We can't have one for both, and introducing a fourth value (null) would only make more confusing for people.

Piinks commented:
> it doesn't work

We are still disagreeing though. 

Can you consider that the user will change the value of `strokeCap` based on the `value` they they are providing to the progress indicator?

I feel we are at an impasse here and unfortunately I cannot spend too much more time on this PR.
The original author was going to reopen their PR before this took it over, I feel we should consider their work and the request that was being made in the first place.

bernaferrari commented:
> The original author was going to reopen their PR before this took it over

The original PR never worked because he didn't notice the code overrides the strokeCap. The only way it would work would be with `null`, but then you have 3 values for similar things.

> Can you consider that the user will change the value of strokeCap based on the value they they are providing to the progress indicator?

I can also consider `butt` is not the opposite of `round`, so if they are playing with it and set to "square", they are introducing a visual bug to their codebase and might only notice in a later date.

bernaferrari committed:
Add `indicatorBorderRadius`.

Remove from theme.
Add `backgroundBorderRadius`.

Use `ShapeDecoration` for `LinearProgressIndicator`.
Fix PR review.

Use `shape` instead of `backgroundBorderRadius`.

bernaferrari committed:
Make strokeCap null.

bernaferrari commented:
I just pushed the strokeCap with 4 states.

bernaferrari committed:
Apply HixieGPT.

bernaferrari committed:
Fix test.

bernaferrari committed:
Fix tests.

bernaferrari committed:
Fix analyzer.

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -396,6 +396,7 @@ class _CircularProgressIndicatorPainter extends CustomPainter {
     required this.offsetValue,
     required this.rotationValue,
     required this.strokeWidth,
+    required this.strokeCap,
```
Does this need to required if it is nullable?

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -434,8 +437,12 @@ class _CircularProgressIndicatorPainter extends CustomPainter {
       paint.strokeCap = StrokeCap.square;
+    } else {
+      // Butt when determinate (value != null) && strokeCap == null;
+      paint.strokeCap = strokeCap ?? StrokeCap.butt;
```
`StrokeCap.butt` was not here before. Why this change?

bernaferrari replied:
it wasn't set because it was the default value, we must explicit set it now

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -542,6 +552,26 @@ class CircularProgressIndicator extends ProgressIndicator {
+  ///
+  /// Determines the appearance of the stroke ends in the progress indicator.
+  /// By default, [strokeCap] is null, causing the stroke ends to be set to
+  /// [StrokeCap.square] when [value] is null (indeterminate), and to
+  /// [StrokeCap.butt] when [value] is not null (determinate).
```
What does determinate and indeterminate mean here for folks who may not know?

bernaferrari replied:
The class already provides it, but I can write it better. Do you have a suggestion? 
I'll try `determinate progress indicator`.
<img width="743" alt="image" src="https://user-images.githubusercontent.com/351125/230979573-14d4cc20-5812-402e-a6ba-cf0106c3cd84.png">

bernaferrari commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -542,6 +552,26 @@ class CircularProgressIndicator extends ProgressIndicator {
+  ///
+  /// Using [strokeCap] as [StrokeCap.square] when [value] is not null leads to
+  /// an unconventional display of [value], starting and ending slightly
+  /// off the expected angles. While not recommended, this behavior can be
```
I don't think it is recommended because that's simply not how CircularProgressIndicator was intended to be used (based on the way it dances with butt and square for different scenarios). I think it is fair to warn the users it is probably not what they want (this will suit 0.00001% of them). ShortcutActivator doesn't recommend certain pattern because a different one "closely resambles that of typical platforms". I think we are able to not recommend something that.

 But I can try better wording. What do you think of this?
<img width="695" alt="image" src="https://user-images.githubusercontent.com/351125/230980950-1134dfa6-c0df-4ed2-8789-67cebc9ed4e0.png">

bernaferrari replied:
GPT-4 provided:
> Maybe. I tried with GPT-4, it said:
> Be careful when using [StrokeCap.square] with a non-null [value]. The start and end angles may not be what you expect. For example, with a [value] of 0.5, you might expect the start and end angles to be 90 and 270 degrees. However, the indicator could begin at 85 degrees and end at 275 degrees. This doesn't follow the usual design rules, so use it with caution.

bernaferrari replied:
I liked the lighter, I'll use it. Let me know if you feel it is better.

bernaferrari committed:
Fix PR

bernaferrari committed:
Try GPT-4 generated wording.

bernaferrari committed:
Reword again.

Piinks commented:
We currently cannot consider a language model to have signed the CLA, and so we cannot accept changes written by a language model. Can you remove what was written by the LLM?

bernaferrari committed:
No GPT.

bernaferrari commented:
Here it goes again.

bernaferrari committed:
Fix typo.

bernaferrari commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -542,6 +552,27 @@ class CircularProgressIndicator extends ProgressIndicator {
+  /// Setting [strokeCap] to [StrokeCap.square] with [value] != null will
+  /// result in a different display of [value]. The indicator will start
+  /// drawing from slightly less than the start, and end slightly after
+  /// the end. This might produce unintended results, as the expected behavior
```
I have to say I don't have a single use case for this. But I'm willing to make this sacrifice :)

Piinks replied:
> I don't have a single use case for this

You are not the only one who will be using it. :)
Thanks for updating!

bernaferrari committed:
Fix PR comment.

Piinks commented on packages/flutter/lib/src/material/progress_indicator.dart:
```diff
@@ -542,6 +552,27 @@ class CircularProgressIndicator extends ProgressIndicator {
+  /// Setting [strokeCap] to [StrokeCap.square] with [value] != null will
+  /// result in a different display of [value]. The indicator will start
+  /// drawing from slightly less than the start, and end slightly after
+  /// the end. the end. This will produce an alternative result, as the
```
Typo

```suggestion
  /// the end. This will produce an alternative result, as the
```

Piinks committed:
Update packages/flutter/lib/src/material/progress_indicator.dart

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/169392212-02bb74fe-c6b9-4eb7-aa62-6ddb43a70d5b.png)

rayliverified commented:
Amazing, thank you @Piinks and @bernaferrari very much! 

Kate, can you shed some light on why this PR was accepted and the previous ones were rejected?

bernaferrari commented:
The previous one didn't pass flutter tests.

jhonatan-3a commented:
merge, when will be this in stable ?

bernaferrari commented:
Probably in 2 months

fingerart commented:
May I ask which version this PR was released in?

Hari-07 commented:
@fingerart It's out in [3.13](https://docs.flutter.dev/release/release-notes/release-notes-3.13.0#:~:text=Add%20StrokeCap%20to%20CircularProgressIndicator%20by%20%40bernaferrari%20in%20122664)
